api/db/changelog/db.changelog-29-admin-logging.xml

```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<databaseChangeLog
+xmlns="http://www.liquibase.org/xml/ns/dbchangelog/1.9"
+xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog/1.9
+                    http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-1.9.xsd">
+  <changeSet author="rsasch" id="changelog-29">
+    <createTable tableName="admin_action_history">
+      <column name="history_id" type="bigint" autoIncrement="true">
+        <constraints primaryKey="true" nullable="false"/>
+      </column>
+      <column name="user_id" type="bigint">
+        <constraints nullable="false"/>
+      </column>
+      <column name="target_id" type="bigint">
```

Comment:
not always a user; so far, it can also be a workspace ID

Reply:
Ah, ok. Perhaps there should be multiple nullable columns here then - target_user_id, target_workspace_id, etc? Worried about a polymorphic ID here; would especially be a problem if the ID type wasn't compatible with bigint.

Reply:
Hm...I understand the wariness about compatibility with other IDs, but I also don't mind the idea of enforcing that all DB entities have unique bigint IDs.

Reply:
That's really a secondary concern, it seems like it would be very difficult to use this column programmatically to look something up, since you would have no idea which table to look in without parsing a free text log.

Reply:
Since we don't yet have a clear use case on how this data will be queried/used, I tried to make the data structure as simple as possible.  If it turns out we want this to be explicitly and efficiently searched, I'd definitely go for something more generic (see the two table proposal in https://precisionmedicineinitiative.atlassian.net/browse/RW-377).

Reply:
In general I'd say it's probably prudent to favor a more general representation in response to murky product requirements. For example, it seems probable someone might request that for the admin history (CLI or UI - whereever we use this), we display the target user email instead of an opaque ID. With the current PR, we'd need to do a DB backfill to implement that (or have some very hacky string matching logic).

Possible solutions are: adding another column like "target_type", or switching to multiple nullable columns as I described above.

Anyways, what you have now seems OK if you feel strongly. I just have a sense that storing this information initially may avoid some future work.

Reply:
I believe this is missing a data integrity check. It seems you could add a workspace ID or a user ID here that is invalid. If you wish to keep the generic format, I'd favor CH's suggestion of a "target_type" which you'd then have to use in a manually-defined constraint checking function. Separating into two nullable columns will likely make the check simpler.